Fixes #36035 - Add empty content states for CV repositories

jeremylenz opened the pull request:
#### What are the changes introduced in this pull request?

This PR improves empty states in the new CV UI for ContentViewRepositories, in the following two situations:

1. When your content view doesn't have any repositories added;
2. When there aren't any repositories in your organization.

It also changes the default view from "All" repositories to "Added," and adds call-to-action buttons & links:

1. Show repositories - switches the view from Added to All
2. Enable Red Hat repositories - link
3. Create a custom product - link

#### Considerations taken when implementing this change?

The "Show repositories" button was originally "Add repositories," but I thought it would be confusing to have two different "Add Repositories" buttons on the screen.

#### What are the testing steps for this pull request?

Test the following scenarios:

1. Empty content view with no repositories added
2. Empty content view in an empty organization with no products or Red Hat repos enabled
3. also make sure I didn't break empty states anywhere else in CV UI or the new host details page

jeremylenz commented:
![image](https://user-images.githubusercontent.com/22042343/223197242-02c30302-7649-4939-a420-f9d1cf49ad6b.png)

jeremylenz commented on webpack/scenes/ContentViews/Details/Repositories/ContentViewRepositories.js:
```diff
@@ -217,8 +231,37 @@ const ContentViewRepositories = ({ cvId, details }) => {
+    </Button>
+  );
+
+  const secondaryActionLink = trulyEmpty ? ('/products/') : undefined;
```
Should this go to `/products/` or `/products/new`? I changed it to products because what if you've created a product already but haven't added repos to it. But I'm on the fence.

sjha4 replied:
`/products/` is good. It has it's own empty state in bastion.. 😝

jeremylenz commented on webpack/scenes/ContentViews/Details/Repositories/ContentViewRepositories.js:
```diff
@@ -230,6 +273,8 @@ const ContentViewRepositories = ({ cvId, details }) => {
     </DropdownItem>,
   ];
 
+  const loadingStatus = (isLoading || status === STATUS.PENDING) ? STATUS.PENDING : status;
```
When you clicked the Show Repositories button, it would flash momentarily with the trulyEmpty state before you saw the loading spinner. This prevents that.

jeremylenz commented on webpack/components/Table/EmptyStateMessage.js:
```diff
@@ -97,7 +97,8 @@ const EmptyStateMessage = ({
         }
 
-        {(showSecondaryActionButton || searchIsActive || !!filtersAreActive) &&
+        {(!showSecondaryActionAnchor &&
```
Added this so that it only shows one secondary action - either the <a> link style or the button with onClick - but not both.

jeremylenz commented on webpack/components/Table/MainTable.js:
```diff
@@ -76,7 +76,7 @@ const MainTable = ({
         body={emptyContentBody}
         happy={happyEmptyContent}
-        search={!happyEmptyContent}
+        search={!happyEmptyContent && !emptyContentOverride}
```
had to add this so that it would show the plus icon and not the search icon

jeremylenz commented:
@sjha4 Updated:

1. Shortened wording
2. Updated empty state as per discussion with Maria
3. Updated API endpoints so that we can show the "trulyEmpty" state immediately.

![image](https://user-images.githubusercontent.com/22042343/223816944-b278e8da-702e-4b6e-b5bc-20799d06640f.png)

jeremylenz commented on webpack/scenes/ContentViews/Details/Repositories/ContentViewRepositories.js:
```diff
@@ -217,8 +247,35 @@ const ContentViewRepositories = ({ cvId, details }) => {
+    </Button>
+  );
+
+  const secondaryActionLink = trulyEmpty ? ('https://docs.theforeman.org/nightly/Managing_Content/index-katello.html#Products_and_Repositories_content-management') : undefined;
```
I'm not sure of the correct way to do this documentation link.

sjha4 commented on webpack/utils/useKatelloDocUrl.js:
Can you explain to me how the downstream links will work? Does the branding change the root URL and the version automatically?

jeremylenz replied:
Yes, the downstream branding plugin should do the requisite magic. I'll ping you with more details..

sjha4 commented:
Works well..

Documentation link is formed correctly but picks up next foreman version 3.7 documentation for katello. We would need to go back a version but I guess it's only a problem for folks running master. Released versions should be fine.

A small naming change as discussed and this should be good to go. ✔️

sjha4 reviewed (approved):
APJ 👍🏼

ekohl commented on webpack/utils/useKatelloDocUrl.js:
```diff
@@ -0,0 +1,18 @@
+// https://docs.theforeman.org/3.7/Managing_Content/index-katello.html#Products_and_Repositories_content-management
+export const useKatelloDocUrl = (guide = 'Managing_Content', hash = '') => {
+  // in dev you'll have to replace the Foreman version in the url with the latest published one
+  const foremanVersion = useForemanVersion();
+  const rootUrl = `https://docs.theforeman.org/${foremanVersion}/`;
+
+  const section = `${guide}/index-katello.html${hash}`;
```
This doesn't belong in the frontend. By having it in the LinksController we can apply branding, which is crucial for Red Hat to properly supports its customers. Note that Red Hat also uses a different file name and directory structure.

ekohl reviewed (commented):
Can we revert the `useKatelloDocUrl` part so we don't start to rely on it? That should be done before it's branched for the next release, or the work within Red Hat downstream will be a lot harder.

jeremylenz commented:
@ekohl This won't be in released code until Katello 4.9, so we have some time. It does go through the links controller (since `foremanUrl` leads there.) If you want to raise a PR doing it a more correct way (with the corresponding Foreman changes) I'll be happy to review and make sure they get in.

ekohl commented:
From now on there are less than 3 weeks left until we branch. In my experience that time flies by. And I think it's also a lot of wasted time to introduce temporary API while it's clear we need a permanent one.

ekohl commented:
Just as I predicted, that this was never picked up and now Katello 4.9 is GA. With my downstream hat on, this is going to require special attention so downstream products don't link to upstream resources.

ekohl commented:
Sadly nothing happened with this, but I've just opened https://github.com/Katello/katello/pull/11061 to fix it.
